pkg/controllers/dashboardapi/settings/settings_test.go

```diff
@@ -257,9 +250,52 @@ func populateTestCases() []*testCase {
 		}
 		test.existingSetting.Name = settingName
 	}
+
 	return testCases
 }
 
-func stringPtr(val string) *string {
-	return &val
+func TestMarkSettingAsUnknownRetry(t *testing.T) {
+	store := map[string]v3.Setting{
+		"unknown": {
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "unknown",
+			},
+			Value:   "unknown",
+			Default: "unknown",
+		},
+	}
+
+	client := fake.NewMockNonNamespacedControllerInterface[*v3.Setting, *v3.SettingList](gomock.NewController(t))
+
+	client.EXPECT().Get(gomock.Any(), gomock.Any()).DoAndReturn(func(name string, options metav1.GetOptions) (*v3.Setting, error) {
+		val, ok := store[name]
+		if !ok {
+			return nil, apierrors.NewNotFound(schema.GroupResource{}, name)
+		}
+
+		return &val, nil
+	}).Times(1)
+
+	var updateRun int
+	client.EXPECT().Update(gomock.Any()).DoAndReturn(func(s *v3.Setting) (*v3.Setting, error) {
+		defer func() { updateRun++ }()
+
+		if updateRun == 0 { // Fail the the first update to force retry.
+			return nil, apierrors.NewConflict(schema.GroupResource{}, s.Name, fmt.Errorf("some error"))
+		}
+
+		store[s.Name] = *s
+		return s, nil
+	}).Times(2)
+
+	provider := settingsProvider{
+		settings: client,
+	}
+
+	unknown := store["unknown"]
+
+	err := provider.markSettingAsUnknown(&unknown)
```

Comment:
Nit: Generally speaking, you want to test the exported interface when possible since that's what other packages can consume. 

I'm marking this as a nit due to my above recommendation that this be an exported function, and because this is a new function, so we have been known to make exceptions when you aren't touching an existing function.

Reply:
> Generally speaking, you want to test the exported interface when possible since that's what other packages can consume.
Not if you want to test an essential behavior that is not part of the exported API as in this case.

Reply:
I would argue that this is part of the exported API, since this is part of the behavior that you expect when you call `SetAll

Reply:
And in the end I would agree 😄. Testing a narrower aspect in isolation can be beneficial though. But at the same time I see no reason why we can't test it through the call of the exported method especially in this particular case. The test is updated.